sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReader.java

```diff
@@ -0,0 +1,148 @@
+  private Iterator<Row> rowIterator;
+  private final ReadRowsRequest request;
+  private SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private BoundedSource<T> source;
+  private BigQueryOptions options;
+  private Row currentRow;
+
+  private BigQueryParallelReader(
+      Session session,
+      ReadLocation readLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    this.session = checkNotNull(session, "session");
+    this.client = checkNotNull(client, "client");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.source = checkNotNull(source, "source");
+    this.options = options;
+
+    ReadRowsRequest.Builder builder = ReadRowsRequest.newBuilder().setReadLocation(readLocation);
+    if (rowBatchSize != null) {
+      builder.setOptions(ReadOptions.newBuilder().setMaxRows(rowBatchSize).build());
+    }
+
+    this.request = builder.build();
+  }
+
+  public static <T> BigQueryParallelReader<T> create(
+      Session session,
+      ReadLocation initialReadLocation,
+      Integer rowBatchSize,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      BoundedSource<T> source,
+      BigQueryServices client,
+      BigQueryOptions options) {
+    return new BigQueryParallelReader<>(
+        session,
+        initialReadLocation,
+        rowBatchSize,
+        parseFn,
+        source,
+        client,
+        options);
+  }
+  /**
+   * Empty operation, the table is already open for read.
+   * @throws IOException on failure
+   */
+  @Override
+  public boolean start() throws IOException {
+    responseIterator = client.getTableReadService(options).readRows(request);
+    if (responseIterator != null && responseIterator.hasNext()) {
+      rowIterator = responseIterator.next().getRowsList().iterator();
+      if (rowIterator.hasNext()) {
+        currentRow = rowIterator.next();
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    if (responseIterator == null) {
+      throw new IOException("start() needs to be called first");
+    }
+    if (rowIterator != null && rowIterator.hasNext()) {
+      currentRow = rowIterator.next();
+      return true;
+    }
+    if (rowIterator == null || !rowIterator.hasNext()) {
+      if (responseIterator.hasNext()) {
+        rowIterator = responseIterator.next().getRowsList().iterator();
+        if (rowIterator.hasNext()) {
+          currentRow = rowIterator.next();
+          return true;
+        }
+      }
+    }
+    currentRow = null;
+    return false;
+  }
+
+  @Override
+  public T getCurrent() {
+    if (currentRow == null) {
+      return null;
+    }
+    return parseFn.apply(new SchemaAndRowProto(session.getProjectedSchema(), currentRow));
+  }
+
+  @Override
+  public void close() {
+    // How to close a stream from client?
```

Comment:
Please clarify this comment: is there currently a resource leak in the code, or is this harmless?

Reply:
This is harmless. I've updated the code comment.